vxhunter/vxutility/symbol_table.py

```diff
@@ -110,56 +104,47 @@ def parse_sym(offset, sym_st_fmt, sym_size, sym_types):
     return sym_dict(offset, *fields), is_sym_valid(sym_types, *fields)
 
 
-def get_symtab_bounds(blk, vx_ver, symbol_fn):
-    '''
-    Try to get the VxWorks 5, 6, or 7 symbol table.
-    '''
+def fix_image_base(blk, vx_ver):
+    """Note: this may seem redundant to duplicate effort, trying to find symtable here and then again
+    in get_symtab_bounds(). However rebasing is needed
+    """
     sym_st_fmt, sym_st_size = get_symbol_fmt(endy_str, word_str, vx_ver)
     sym_types = SYM_TYPES[vx_ver]
 
-    symtab = []
-    i = blk.start.offset
-
-    while i < blk.end.offset - sym_st_size:
-        sym, sym_valid = parse_sym(i, sym_st_fmt, sym_st_size, sym_types)
-
-        # if the current symbol struct is not a valid symbol, only increment the cursor by `word_size` bytes
-        # in case we are unaligned with the symbol table (this assumes the symbol table is word-aligned)
-        if sym is None or not sym_valid:
-            symtab = []
-            i += word_size
-            continue
-
-        name_addr = sym['name_addr']
+    left_off, symtab = find_sym_table(blk, blk.start.offset, sym_st_fmt, sym_st_size, sym_types, vx_ver)
+    if len(symtab) < SYMTAB_MIN_COUNT:
+        # If no matches were found it could be that they were thrown out because the offset is so large
+        # that none of the syms match fall in the memory range of the image.
+        # Try again, this time without verifying the addresses.
+        left_off, symtab = find_sym_table(blk, blk.start.offset, sym_st_fmt, sym_st_size, sym_types, vx_ver,
+                                          verify_addresses=False)
 
-        # the name pointer must be within the bounds of the memory block
-        #if name_addr < blk.start.offset or name_addr >= blk.end.offset:
-        if not is_offset_in_current_program(name_addr):
-            symtab = []
-            i += word_size
-            continue
+        # throw out the candidate symbol table and return if it's below the threshold count
+        if len(symtab) < SYMTAB_MIN_COUNT:
+            return None
 
-        sym_type = sym['flag']
+    #for sym in symtab:
+    #    print_out("name 0x%08x" %sym['name_addr'])
 
-        # if the symbol type resides in the .text or .data section, make sure it's dest pointer is within the block
-        if sym_type in TEXT_SYM_TYPES[vx_ver] or sym_type in DATA_SYM_TYPES[vx_ver]:
-            dest_addr = sym['dest_addr']
+    symtab = sorted(symtab, key=lambda x: x['offset'])
 
-            #if dest_addr < blk.start.offset or dest_addr >= blk.end.offset:
-            if not is_offset_in_current_program(dest_addr):
-                symtab = []
-                i += word_size
-                continue
+    # if we have a block at offset zero, it's probably not correct...
```

Comment:
why?

Reply:
because at this point in the code other things have been tried and failed. I guess my other thought process was regardless of the device, you usually want memory location zero (null) to be an invalid memory location. If you have a null pointer, the code should segfault when trying to access it and shouldn't return valid memory.

Additionally If the image already has a memory base set, we shouldn't try to rebase it, and if it's a monolithic image that was imported the base will be zero (at least in all the imports I did in Ghidra) but i may be basing this on flawed understanding, so happy to change it.

I also found only one Vxworks image that has a base of zero where it's valid, and if it is valid then the previous steps will have worked.

Reply:
OK. I think that checks out. A lot of our images do have a base of 0 (ARM EVT). But I'm fine using this as a proxy for the base address not having been set since if it were good it would have succeeded earlier.